File: frontend/lib/api/follow.ts
/**
 * Follow API Client
 * 关注相关的 API 客户端
 */

import { apiClient } from './client';

export interface FollowStats {
  followers_count: number;
  following_count: number;
}

export interface FollowStatus {
  is_following: boolean;
  is_followed_by: boolean;
}

export interface FollowerInfo {
  id: number;
  user_id: number;
  username: string;
  full_name: string | null;
  avatar_url: string | null;
  bio: string | null;
  followed_at: string;
}

export interface FollowersListResponse {
  followers: FollowerInfo[];
  total: number;
  page: number;
  page_size: number;
}

export interface FollowingListResponse {
  following: FollowerInfo[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * 关注 API 客户端
 */
export const followApi = {
  /**
   * 关注用户
   */
  async followUser(userId: number): Promise<void> {
    await apiClient.post(`/follows/follow/${userId}`);
  },

  /**
   * 取消关注用户
   */
  async unfollowUser(userId: number): Promise<void> {
    await apiClient.delete(`/follows/unfollow/${userId}`);
  },

  /**
   * 检查关注状态
   */
  async checkFollowStatus(userId: number): Promise<FollowStatus> {
    return await apiClient.get<FollowStatus>(`/follows/check/${userId}`);
  },

  /**
   * 获取粉丝列表
   */
  async getFollowers(
    userId: number,
    page: number = 1,
    pageSize: number = 20
  ): Promise<FollowersListResponse> {
    return await apiClient.get<FollowersListResponse>(
      `/follows/followers/${userId}`,
      { params: { page, page_size: pageSize } }
    );
  },

  /**
   * 获取关注列表
   */
  async getFollowing(
    userId: number,
    page: number = 1,
    pageSize: number = 20
  ): Promise<FollowingListResponse> {
    return await apiClient.get<FollowingListResponse>(
      `/follows/following/${userId}`,
      { params: { page, page_size: pageSize } }
    );
  },

  /**
   * 获取关注统计
   */
  async getFollowStats(userId: number): Promise<FollowStats> {
    return await apiClient.get<FollowStats>(`/follows/stats/${userId}`);
  },

  /**
   * 获取当前用户的关注列表
   */
  async getMyFollowing(
    page: number = 1,
    pageSize: number = 20
  ): Promise<FollowingListResponse> {
    return await apiClient.get<FollowingListResponse>('/follows/me/following', {
      params: { page, page_size: pageSize },
    });
  },

  /**
   * 获取当前用户的粉丝列表
   */
  async getMyFollowers(
    page: number = 1,
    pageSize: number = 20
  ): Promise<FollowersListResponse> {
    return await apiClient.get<FollowersListResponse>('/follows/me/followers', {
      params: { page, page_size: pageSize },
    });
  },
};

File: backend/app/schemas/follow.py
"""
Follow Schemas
关注相关的数据验证模型
"""

from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


class FollowBase(BaseModel):
    """关注基础模型"""
    follower_id: int
    following_id: int


class FollowCreate(FollowBase):
    """创建关注请求"""
    pass


class FollowResponse(FollowBase):
    """关注响应"""
    id: int
    followed_at: datetime
    follower_username: Optional[str] = None
    following_username: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class FollowerInfo(BaseModel):
    """粉丝信息"""
    id: int
    user_id: int
    username: str
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    bio: Optional[str] = None
    followed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FollowingInfo(BaseModel):
    """关注信息"""
    id: int
    user_id: int
    username: str
    full_name: Optional[str] = None
    avatar_url: Optional[str] = None
    bio: Optional[str] = None
    followed_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FollowersListResponse(BaseModel):
    """粉丝列表响应"""
    followers: List[FollowerInfo]
    total: int
    page: int
    page_size: int


class FollowingListResponse(BaseModel):
    """关注列表响应"""
    following: List[FollowingInfo]
    total: int
    page: int
    page_size: int


class FollowStatsResponse(BaseModel):
    """关注统计响应"""
    followers_count: int
    following_count: int


class FollowStatusResponse(BaseModel):
    """关注状态响应"""
    is_following: bool
    is_followed_by: bool


class BulkFollowRequest(BaseModel):
    """批量关注请求"""
    user_ids: List[int]


class BulkFollowResponse(BaseModel):
    """批量关注响应"""
    successful: List[int]
    failed: List[dict]
